src/store/Disks.cc

```diff
@@ -487,14 +487,30 @@ Store::Disks::sync()
 
 void
 Store::Disks::markForUnlink(StoreEntry &e) {
-    if (e.swap_filen >= 0)
+    if (e.hasDisk())
         store(e.swap_dirn)->markForUnlink(e);
 }
 
 void
-Store::Disks::unlink(StoreEntry &e) {
-    if (e.swap_filen >= 0)
-        store(e.swap_dirn)->unlink(e);
+Store::Disks::unlinkByKeyIfFound(const cache_key *key)
+{
+    for (int i = 0; i < Config.cacheSwap.n_configured; ++i) {
+        if (dir(i).active())
+            dir(i).unlinkByKeyIfFound(key);
+    }
+}
+
+void
+Store::Disks::unlink(StoreEntry &e)
+{
+    for (int i = 0; i < Config.cacheSwap.n_configured; ++i) {
+        if (dir(i).active()) {
+            if (e.hasDisk(i))
+                dir(i).unlink(e);
+            else
+                dir(i).unlinkByKeyIfFound(reinterpret_cast<const cache_key*>(e.key));
```

Comment:
As already discussed elsewhere, with "approach 2", this inner if-statement would be replaced with just `dir(i).unlink(e);` and each cache dir will do either mark the (locked) entry for unlinking or unlink by key.

Reply:
Done at a40bbf6.
Following it, I removed a hasDisk() check from storeSwapOutFileClosed() because the surrounding code assumes that the entry should be attached to a disk: it uses e->disk() without check. Probably there should be an assert(e->hasDisk()) near existing assert(e->swappingOut())?

Reply:
I would not add a hasDisk() assert to protect disk() calls because those calls have that assert(). 

Since storeSwapOutFileClosed() does not manipulate raw disk indexes/pointers itself (only dumps them in stale debugs() messages), I think we are OK without an assert protecting such raw manipulations.

I consider this change request addressed.